packages/cozy-realtime/src/index.js

```diff
@@ -99,228 +120,233 @@ const configTypes = {
+        listener(payload.doc)
+      })
     }
+  }
 
-    try {
-      socket = await connectWebSocket(
-        config,
-        onSocketMessage,
-        onSocketClose,
-        NUM_RETRIES,
-        RETRY_BASE_DELAY
+  const onSocketClose = (event, numRetries, retryDelay) => {
+    if (!event.wasClean) {
+      console.warn(
+        `WebSocket closed unexpectedly with code ${event.code} and ${
+          event.reason ? `reason: '${event.reason}'` : 'no reason'
+        }.`
       )
-    } catch (error) {
-      reject(error)
-    }
 
-    cozySocket = {
-      subscribe: (doctype, event, listener) => {
-        if (typeof listener !== 'function')
-          throw new Error('Realtime event listener must be a function')
-
-        if (!listeners[doctype]) {
-          listeners[doctype] = {}
-          subscribeWhenReady(doctype, socket)
-        }
-
-        listeners[doctype][event] = (listeners[doctype][event] || []).concat([
-          listener
-        ])
-      },
-      unsubscribe: (doctype, event, listener) => {
-        if (
-          listeners[doctype] &&
-          listeners[doctype][event] &&
-          listeners[doctype][event].includes(listener)
-        ) {
-          listeners[doctype][event] = listeners[doctype][event].filter(
-            l => l !== listener
-          )
-        }
+      if (numRetries) {
+        console.warn(`Reconnecting ... ${numRetries} tries left.`)
+        setTimeout(() => {
+          try {
+            socket = connectWebSocket(
+              config,
+              onSocketMessage,
+              onSocketClose,
+              --numRetries,
+              retryDelay + 1000,
+              true
+            )
+          } catch (error) {
+            console.error(
+              `Unable to reconnect to realtime. Error: ${error.message}`
+            )
+          }
+        }, retryDelay)
       }
     }
+  }
 
-    resolve(cozySocket)
-  })
-}
+  try {
+    socket = connectWebSocket(
+      config,
+      onSocketMessage,
+      onSocketClose,
+      NUM_RETRIES,
+      RETRY_BASE_DELAY
+    )
+  } catch (error) {
+    throw error
+  }
+
+  return {
+    subscribe: (doctype, event, listener, docId) => {
+      if (typeof listener !== 'function')
+        throw new Error('Realtime event listener must be a function')
 
-// Returns the Promise of a subscription to a given doctype and document
-export function subscribe(config, doctype, doc, parse = doc => doc) {
-  return subscribeAll(config, doctype, parse).then(subscription => {
-    // We will call the listener only for the given document, so let's curry it
-    const docListenerCurried = listener => {
-      return syncedDoc => {
-        if (syncedDoc._id === doc._id) {
-          listener(syncedDoc)
-        }
+      const listenerKey = getListenerKey(doctype, docId)
+
+      if (!listeners.has(listenerKey)) {
+        listeners.set(listenerKey, {})
+        subscribeWhenReady(doctype, socket, docId)
```

Comment:
If I understand correctly, if we setup a listener for `doctype=io.cozy.jobs & docId= 123` for both `updated` and `deleted` events, we will call `subscribeWhenReady` twice, and the stack will receive 2 times the message to subscribe for this document. Right?